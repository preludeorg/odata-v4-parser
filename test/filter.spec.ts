import { get } from '@newdash/newdash';
import { Edm } from '@odata/metadata';
import { defaultParser, ODataFilter, ODataParam } from '../src';


describe('Filter Test Suite', () => {

  it('should support simple eq', () => {
    defaultParser.query('$format=json&$filter=A eq 2');
  });

  it('should support complex query', () => {
    defaultParser.query('$format=json&$filter=(A eq 2) and (B eq 3 or B eq 4)');
  });

  it('should support eq with guid & string', () => {
    const ast = defaultParser.query("$filter=A eq 702dac82-923d-4958-805b-ca41c593d74f and B eq 'strValue'");
    expect(ast).not.toBeNull();
    expect(get(ast, 'value.options[0].value.value.left.value.right.value')).toBe(Edm.Guid.className);
    expect(get(ast, 'value.options[0].value.value.right.value.right.value')).toBe(Edm.String.className);
  });
  
  it('shuold support filter without double quote', () => {
    const filter = ODataFilter.New().field("key").eq('val')
    const filterStr = ODataParam.New().filter(filter).toString()
    expect(filterStr).toMatchSnapshot()
    const ast = defaultParser.query(filterStr)
    expect(ast).toMatchSnapshot()
  })

  it('shuold support filter with double quote', () => {
    const filter = ODataFilter.New().field("key").eq('val"')
    const filterStr = ODataParam.New().filter(filter).toString()
    expect(filterStr).toMatchSnapshot()
    const ast = defaultParser.query(filterStr)
    expect(ast).toMatchSnapshot()
  })

  it('should support complex filter', () => {

    const sFilter = ODataFilter.New()
      .field('A').eq(1).field('A').eq(2)
      .field('B').gt(3)
      .field('C').between(1, 3)
      .field('F').between(1, 3, true)
      .field('E').in(['a', 'c', 'd'])
      .field('year(Date)').eq(2010)
      .field('Date2').gt(Edm.DateTimeOffset.createValue('2020-07-30T03:16:27.023Z'))
      .field('Date3').lt(Edm.DateTimeOffset.createValue(new Date()))
      .toString();

    defaultParser.filter(sFilter);

  });

});
