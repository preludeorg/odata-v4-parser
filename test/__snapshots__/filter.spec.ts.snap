// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter Test Suite shuold support filter with double quote 1`] = `"$filter=key eq 'val"'"`;

exports[`Filter Test Suite shuold support filter with double quote 2`] = `
Token {
  "next": 21,
  "position": 0,
  "raw": "$filter=key eq 'val"'",
  "type": "QueryOptions",
  "value": {
    "options": [
      Token {
        "next": 21,
        "position": 0,
        "raw": "$filter=key eq 'val"'",
        "type": "Filter",
        "value": Token {
          "next": 21,
          "position": 8,
          "raw": "key eq 'val"'",
          "type": "EqualsExpression",
          "value": {
            "left": Token {
              "next": 11,
              "position": 8,
              "raw": "key",
              "type": "FirstMemberExpression",
              "value": Token {
                "next": 11,
                "position": 8,
                "raw": "key",
                "type": "MemberExpression",
                "value": Token {
                  "next": 11,
                  "position": 8,
                  "raw": "key",
                  "type": "PropertyPathExpression",
                  "value": Token {
                    "next": 11,
                    "position": 8,
                    "raw": "key",
                    "type": "ODataIdentifier",
                    "value": {
                      "name": "key",
                    },
                  },
                },
              },
            },
            "right": Token {
              "next": 21,
              "position": 15,
              "raw": "'val"'",
              "type": "Literal",
              "value": "Edm.String",
            },
          },
        },
      },
    ],
  },
}
`;

exports[`Filter Test Suite shuold support filter with other chars 1`] = `"$filter=key eq 'val$%^&*()_'"`;

exports[`Filter Test Suite shuold support filter with other chars 2`] = `
Token {
  "next": 28,
  "position": 0,
  "raw": "$filter=key eq 'val$%^&*()_'",
  "type": "QueryOptions",
  "value": {
    "options": [
      Token {
        "next": 28,
        "position": 0,
        "raw": "$filter=key eq 'val$%^&*()_'",
        "type": "Filter",
        "value": Token {
          "next": 28,
          "position": 8,
          "raw": "key eq 'val$%^&*()_'",
          "type": "EqualsExpression",
          "value": {
            "left": Token {
              "next": 11,
              "position": 8,
              "raw": "key",
              "type": "FirstMemberExpression",
              "value": Token {
                "next": 11,
                "position": 8,
                "raw": "key",
                "type": "MemberExpression",
                "value": Token {
                  "next": 11,
                  "position": 8,
                  "raw": "key",
                  "type": "PropertyPathExpression",
                  "value": Token {
                    "next": 11,
                    "position": 8,
                    "raw": "key",
                    "type": "ODataIdentifier",
                    "value": {
                      "name": "key",
                    },
                  },
                },
              },
            },
            "right": Token {
              "next": 28,
              "position": 15,
              "raw": "'val$%^&*()_'",
              "type": "Literal",
              "value": "Edm.String",
            },
          },
        },
      },
    ],
  },
}
`;

exports[`Filter Test Suite shuold support filter with single quote 1`] = `"$filter=key eq 'T''A First'"`;

exports[`Filter Test Suite shuold support filter with single quote 2`] = `
Token {
  "next": 27,
  "position": 0,
  "raw": "$filter=key eq 'T''A First'",
  "type": "QueryOptions",
  "value": {
    "options": [
      Token {
        "next": 27,
        "position": 0,
        "raw": "$filter=key eq 'T''A First'",
        "type": "Filter",
        "value": Token {
          "next": 27,
          "position": 8,
          "raw": "key eq 'T''A First'",
          "type": "EqualsExpression",
          "value": {
            "left": Token {
              "next": 11,
              "position": 8,
              "raw": "key",
              "type": "FirstMemberExpression",
              "value": Token {
                "next": 11,
                "position": 8,
                "raw": "key",
                "type": "MemberExpression",
                "value": Token {
                  "next": 11,
                  "position": 8,
                  "raw": "key",
                  "type": "PropertyPathExpression",
                  "value": Token {
                    "next": 11,
                    "position": 8,
                    "raw": "key",
                    "type": "ODataIdentifier",
                    "value": {
                      "name": "key",
                    },
                  },
                },
              },
            },
            "right": Token {
              "next": 27,
              "position": 15,
              "raw": "'T''A First'",
              "type": "Literal",
              "value": "Edm.String",
            },
          },
        },
      },
    ],
  },
}
`;

exports[`Filter Test Suite shuold support filter without double quote 1`] = `"$filter=key eq 'val'"`;

exports[`Filter Test Suite shuold support filter without double quote 2`] = `
Token {
  "next": 20,
  "position": 0,
  "raw": "$filter=key eq 'val'",
  "type": "QueryOptions",
  "value": {
    "options": [
      Token {
        "next": 20,
        "position": 0,
        "raw": "$filter=key eq 'val'",
        "type": "Filter",
        "value": Token {
          "next": 20,
          "position": 8,
          "raw": "key eq 'val'",
          "type": "EqualsExpression",
          "value": {
            "left": Token {
              "next": 11,
              "position": 8,
              "raw": "key",
              "type": "FirstMemberExpression",
              "value": Token {
                "next": 11,
                "position": 8,
                "raw": "key",
                "type": "MemberExpression",
                "value": Token {
                  "next": 11,
                  "position": 8,
                  "raw": "key",
                  "type": "PropertyPathExpression",
                  "value": Token {
                    "next": 11,
                    "position": 8,
                    "raw": "key",
                    "type": "ODataIdentifier",
                    "value": {
                      "name": "key",
                    },
                  },
                },
              },
            },
            "right": Token {
              "next": 20,
              "position": 15,
              "raw": "'val'",
              "type": "Literal",
              "value": "Edm.String",
            },
          },
        },
      },
    ],
  },
}
`;
